File: projects/advent_of_code/solutions/year2025/day01/python/solution.py
#!/usr/bin/env python3
"""Advent of Code 2025 Day 1"""

from projects.advent_of_code.python.aoc import lines, read_file


def part1(input_text: str) -> int:
    pos = 50
    count = 0
    for line in lines(input_text):
        dir, dist = line[0], int(line[1:])
        if dir == "L":
            pos -= dist
        else:
            pos += dist
        pos = pos % 100
        if pos == 0:
            count += 1
    return count


def part2(input_text: str) -> int:
    pos = 50
    count = 0
    for line in lines(input_text):
        dir, dist = line[0], int(line[1:])
        if dir == "R":
            count += (pos + dist) // 100
            pos = (pos + dist) % 100
        else:
            if pos == 0:
                count += dist // 100
            elif dist >= pos:
                count += (dist - pos) // 100 + 1
            pos = (pos - dist) % 100
    return count


if __name__ == "__main__":
    input_text = read_file("advent_of_code/solutions/year2025/day01/input.txt")

    print(f"Part 1: {part1(input_text)}")
    print(f"Part 2: {part2(input_text)}")

File: projects/advent_of_code/solutions/year2025/day01/go/solution.go
package main

import (
	"fmt"

	"github.com/jomcgi/homelab/projects/advent_of_code/pkg/aoc"
)

func main() {
	input, err := aoc.ReadFile("advent_of_code/solutions/year2025/day01/input.txt")
	if err != nil {
		panic(err)
	}

	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}

func part1(input string) int {
	lines := aoc.Lines(input)
	pos := 50
	count := 0

	for _, line := range lines {
		if line == "" {
			continue
		}
		dir := line[0]
		dist := aoc.MustInt(line[1:])

		if dir == 'L' {
			pos -= dist
		} else {
			pos += dist
		}
		pos = aoc.Mod(pos, 100)

		if pos == 0 {
			count++
		}
	}

	return count
}

func part2(input string) int {
	lines := aoc.Lines(input)
	pos := 50
	count := 0

	for _, line := range lines {
		if line == "" {
			continue
		}
		dir := line[0]
		dist := aoc.MustInt(line[1:])

		if dir == 'R' {
			count += (pos + dist) / 100
			pos = (pos + dist) % 100
		} else {
			if pos == 0 {
				count += dist / 100
			} else if dist >= pos {
				count += (dist-pos)/100 + 1
			}
			pos = aoc.Mod(pos-dist, 100)
		}
	}

	return count
}
